### DISCUSSION TITLE
Streaming vs Standard mode advantages

### DISCUSSION THREAD

**Discussion by jamiebullock** [MTG/essentia#1362] [General]
The [documentation](https://essentia.upf.edu/essentia_python_tutorial.html) mentions two advantages of Streaming mode over Standard mode:

- Streaming mode involves writing less code
- Streaming mode uses less memory for large files

Are there any other advantages of either mode vs the other?

**Comment by palonso** [comment]
# Streaming pros
- In Python, streaming can be slightly more efficient since you are not binding intermediate outputs to NumPy types. However, this overhead is very small for most applications.  

# Standard pros
- Standard mode follows an imperative approach, which tends to be more familiar to most users.
- Standard mode allows processing intermediate outputs with 3rd party code, which may be required in some applications.

**Comment by nvssynthesis** [comment]
If I'm not mistaken, another possible advantage of streaming mode is that you can customize the network running procedure in a way that is not possible with standard.
With streaming, once you've got a Network n, by default you'd say
n.run();
in C++, which would call runPrepare() and then runStep() until there are no more tokens to process.
But you could also just manually call runPrepare() and then runStep() until it completes, instead of run(). This allows for injecting other bits of functionality, such as checking in with the progress of the algorithms, early exiting if it's taking too long (e.g. with a huge audio file), etc.
I don't think there is a way to do this with standard, which has the method compute() but no intermediate processes.

**Reply by nolan-veed** [reply]
Yes. This was my biggest hurdle when trying to understand how to integrate the C++ library into our codebase.

I am handling FFmpeg AVFrames as they come in... so I didn't want to store all the frames/samples in memory. So, I am creating a `Network n(gen);` where gen is my new streaming node/algorithm. And then calling `n.runPrepare();` once and then `n.runStep();` in a loop while doing other stuff.

The standard was would take up too much memory I think.
